service/policy/adr/0002-pagination-list-rpcs.md

```diff
@@ -0,0 +1,175 @@
+# Pagination in policy LIST RPCs
+
+## Table of Contents
+
+- [Background](#background)
+- [Chosen Option](#chosen-option)
+- [Considered Options](#considered-options)
+  - [LIMIT + OFFSET](#limit--offset)
+  - [Keyset Pagination](#keyset-pagination)
+  - [Cursor Pagination](#cursor-pagination)
+
+## Background
+
+At present, policy LIST RPCs are completely open-ended.
+
+Attribute Namespaces, Definitions, and Values LIST calls may be filtered by _active_ state.
+
+All Policy objects may be retrieved without quantity limits. This presents a challenge at scale if there
+are a very large number of any policy object in the platform database when responses become overwhelmingly
+large.
+
+Introduction of a `limit` on retrieved items in LIST procedure call responses necessitates the simultaneous introduction of
+pagination. This ADR clarifies the unified approach we will take within policy service LIST RPCs
+and at the database level for this pagination.
+
+## Chosen Option
+
+[LIMIT + OFFSET](#limit--offset)
+
+Because we do not know the likelihood of platforms running with Policy where any individual object has
+enough rows to experience the at-scale performance concerns of `offset` pagination, we will prefer
+this simple implementation for now and leave the door open for cursor-based pagination to solve the performance
+constraint should it be a realized problem in the future.
+
+## Considered Options
+
+### LIMIT + OFFSET
```

Comment:
> Will there be a way to say just give me all items?

This is the current state, and I don't think it's realistic if we're assuming at-scale deployment with policy data we don't control as the platform. I think as a platform there are only limited assumptions we can make about the number of any given policy object that will exist ([Hyrum's Law](https://www.hyrumslaw.com/): With a sufficient number of users of an API, it does not matter what you promise in the contract: all observable behaviors of your system will be depended on by somebody."). If an administrator of a platform decides to add 5,000,000 attributes, it's not feasible for the platform to be expected to bear a `ListAttributes` call that has 5M attributes and all they entail in a single response. 

I think it's a safer approach to enforce a limit now and decide on an approach to satisfy pagination beyond the limit.

That said, with an offset approach, it would be possible to make an SDK function for any policy object like `ListAllAttributes` that opens 5 "workers" (language agnostic) to concurrently fetch all the policy attributes until exhausted and streams the response LIST. With a cursor approach, we would be dependent upon each response's cursor to get the next, so we would be limited in concurrency and the helpfulness of streaming.

> Do we need to add a layer of abstraction to the list operation so we have some type of iterator pattern. This way a user doesn't need to worry about offsets and pages and just could be like .Next()?

This is actually what the `cursor` approach would look like. You can imagine a social media feed with infinite scroll where you can scroll up or down but never "skip" forward/backward 4 pages either direction.

I think there are two use cases for policy LIST calls:

1. administration of platform policy
- "I am looking at a GUI and need to navigate a table or some kind of UX displaying my attributes or subject mappings to C~R~UD them"
- "I know the object I need to see is last on the list, but I need to click/tab X number of times to get there"
- "I don't know where the object I need to see is in the list, but I want to find out and navigate there again fast after I do something else"

2. service-based needs as consumers of policy
- authorization service calling `ListAttributes` to check entitlements
- PEPs/PDPs retrieving policy to drive feature behaviors
- ~the SDKs checking KAS grants~ (not open-ended LIST calls as the value FQNs are known on TDF create)

Reply:
@strantalis does this answer your questions?

Reply:
If you want any kind of abstraction then you could take the HATEOS approach and return the limit and next offset with the response so you can fetch without doing any kind of computation.

Reply:
I think an organization having 5M attributes would be a bit nuts but I didn't expect a moderate size org to have 50k OUs.